nemo/lightning/resume.py

```diff
@@ -1,183 +1,209 @@
+        checkpoint_dir = (
+            Path(self.resume_from_directory) if self.resume_from_directory else Path(Path(log_dir) / "checkpoints")
+        )
+
+        # when using distributed checkpointing, checkpoint_dir is a directory of directories
+        # we check for this here
+        dist_checkpoints = [d for d in list(checkpoint_dir.glob("*")) if d.is_dir()]
+        end_dist_checkpoints = [d for d in dist_checkpoints if d.match("*end")]
+        last_dist_checkpoints = [d for d in dist_checkpoints if d.match("*last")]
+
+        end_checkpoints = _filter_out_unfinished_checkpoints(end_dist_checkpoints)
+        last_checkpoints = _filter_out_unfinished_checkpoints(last_dist_checkpoints)
+
+        if not checkpoint_dir.exists() or (not len(end_checkpoints) > 0 and not len(last_checkpoints) > 0):
+            if self.resume_ignore_no_checkpoint:
+                warn = f"There were no checkpoints found in checkpoint_dir or no checkpoint folder at checkpoint_dir :{checkpoint_dir}. "
+                if checkpoint is None:
+                    warn += "Training from scratch."
+                logging.warning(warn)
+            else:
+                if self.selective_restore_config:
+                    # resume_if_exists is True but run is not resumable. Do not fail and try to do selective restore later instead.
+                    return None
                 else:
                     raise NotFoundError(
                         f"There were no checkpoints found in checkpoint_dir or no checkpoint folder at checkpoint_dir :{checkpoint_dir}. Cannot resume."
                     )
-            elif len(end_checkpoints) > 0:
-                if self.resume_past_end:
-                    if len(end_checkpoints) > 1:
-                        if 'mp_rank' in str(end_checkpoints[0]):
-                            checkpoint = end_checkpoints[0]
-                        else:
-                            raise ValueError(f"Multiple checkpoints {end_checkpoints} that matches *end.ckpt.")
-                else:
-                    raise ValueError(
-                        f"Found {end_checkpoints[0]} indicating that the last training run has already completed."
-                    )
-            elif len(last_checkpoints) > 1:
-                if any([s for s in ['mp_rank', 'tp_rank', 'fsdp_shard'] if s in str(last_checkpoints[0])]):
-                    checkpoint = last_checkpoints[0]
-                    checkpoint = uninject_model_parallel_rank(checkpoint)
-                else:
-                    # Select the checkpoint with the latest modified time
-                    checkpoint = sorted(last_checkpoints, key=lambda pth: pth.lstat().st_mtime, reverse=True)[0]
-                    logging.warning(
-                        f"Multiple checkpoints {last_checkpoints} matches *last.ckpt. Selecting one with the latest modified time."
-                    )
+        elif len(end_checkpoints) > 0:
+            if self.resume_past_end:
```

Comment:
make this `if not self.resume_past_end`, raise the exception and move the nested block one level up.